### PR TITLE
[PropertyInfo] fix resolving self to name of the analyzed class

### LINKED ISSUE #37149: [PropertyInfo] Bug creating the Type for a property 'self' with PHPDocumentor
**Symfony version(s) affected**: All current supported versions

**Description**  
The class name specified if the return type is `self` will become `elf`. However ideally it should return the name of the class that the property belongs to.

**How to reproduce**  
It has been picked up as an issue in API Platform here: https://github.com/api-platform/core/issues/3344 - if there is an annotation `@return self` - the `Type` object returned will define the class as `elf`

**Possible Solution**  
The simplest fix that I can see would be adding it to the list of unsupported return types here:
https://github.com/symfony/property-info/blob/82d9fc8b6a531f479d9545e229f8a5e35dc12775/Util/PhpDocTypeHelper.php#L98

If possible it'd be better to return the FQCN of the object - but I'm not sure how that would work if a patch is applied to also support the return type `static` that looks to be coming in PHP 8.

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #37149
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php [hunk 1]
```diff
@@ -146,7 +146,13 @@ public function getTypes($class, $property, array $context = []): ?array
         /** @var DocBlock\Tags\Var_|DocBlock\Tags\Return_|DocBlock\Tags\Param $tag */
         foreach ($docBlock->getTagsByName($tag) as $tag) {
             if ($tag && !$tag instanceof InvalidTag && null !== $tag->getType()) {
-                $types = array_merge($types, $this->phpDocTypeHelper->getTypes($tag->getType()));
+                foreach ($this->phpDocTypeHelper->getTypes($tag->getType()) as $type) {
+                    if (Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType() && 'self' === $type->getClassName()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]: what about "static" and "parent"?

Comment (xabbuh) [reply src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]: `static` is something that would need to be evaluated at runtime. Not sure how to deal with that here.

Comment (nicolas-grekas) [reply src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]: true, let's leave it as is - only "parent" and "self" can be resolved

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

**Pull request merged**